Samples/Islands/DrawingIsland/DrawingIslandComponents/TextRenderer.cpp

```diff
@@ -0,0 +1,183 @@
+        m_d2dDevice = nullptr;
+        m_compositionGraphicsDevice = nullptr;
+    }
+
+    void TextRenderer::CreateGraphicsDevice()
+    {
+        uint32_t createDeviceFlags = D3D11_CREATE_DEVICE_BGRA_SUPPORT;
+
+        // Array with DirectX hardware feature levels in order of preference.
+        D3D_FEATURE_LEVEL featureLevels[] =
+        {
+            D3D_FEATURE_LEVEL_11_1,
+            D3D_FEATURE_LEVEL_11_0,
+            D3D_FEATURE_LEVEL_10_1,
+            D3D_FEATURE_LEVEL_10_0,
+            D3D_FEATURE_LEVEL_9_3,
+            D3D_FEATURE_LEVEL_9_2,
+            D3D_FEATURE_LEVEL_9_1
+        };
+
+        // Create the Direct3D 11 API device object and a corresponding context.
+        winrt::com_ptr<::ID3D11Device> d3dDevice;
+        winrt::com_ptr<::ID3D11DeviceContext> d3dImmediateContext;
+        D3D_FEATURE_LEVEL d3dFeatureLevel{ D3D_FEATURE_LEVEL_9_1 };
+
+        winrt::check_hresult(
+            ::D3D11CreateDevice(
+                nullptr, // Default adapter.
+                D3D_DRIVER_TYPE_HARDWARE,
+                0, // Not asking for a software driver, so not passing a module to one.
+                createDeviceFlags, // Set debug and Direct2D compatibility flags.
+                featureLevels,
+                ARRAYSIZE(featureLevels),
+                D3D11_SDK_VERSION,
+                d3dDevice.put(),
+                &d3dFeatureLevel,
+                d3dImmediateContext.put()
+            )
+        );
+
+        // Initialize Direct2D resources.
+        D2D1_FACTORY_OPTIONS d2d1FactoryOptions{ D2D1_DEBUG_LEVEL_NONE };
+
+        // Create the Direct2D device object.
+        // Obtain the underlying DXGI device of the Direct3D device.
+        auto dxgiDevice = d3dDevice.as<::IDXGIDevice>();
+
+        winrt::com_ptr<ID2D1Device6> d2dDevice;
+        winrt::check_hresult(
+            m_d2dFactory->CreateDevice(dxgiDevice.get(), d2dDevice.put())
+        );
+
+        // Create the composition graphics device.
+        auto compositorInterop = m_compositor.as<winrt::Microsoft::UI::Composition::ICompositorInterop>();
+        winrt::Microsoft::UI::Composition::ICompositionGraphicsDevice compositionGraphicsDevice;
+        winrt::check_hresult(compositorInterop->CreateGraphicsDevice(d2dDevice.get(), &compositionGraphicsDevice));
+
+        // Save the newly-created objects.
+        m_compositionGraphicsDevice = compositionGraphicsDevice.as<CompositionGraphicsDevice>();
+        m_d2dDevice = std::move(d2dDevice);
+    }
+
+    void TextRenderer::Render(
+        _In_z_ WCHAR const* text,
+        Windows::UI::Color backgroundColor,
+        Windows::UI::Color textColor,
+        SpriteVisual const& visual
+    )
+    {
+        constexpr float marginLeft = 5;
+        constexpr float marginRight = 5;
+        constexpr float marginTop = 5;
+        constexpr float marginBottom = 5;
+
+        winrt::com_ptr<IDWriteTextLayout> textLayout;
+        winrt::check_hresult(m_dwriteFactory->CreateTextLayout(
```

Comment:
Can you add some comments here about why this is doing this? For example, is this how you measure the text so we know how large to make the backing surface?